File: src/effects/Flanger.cpp
#include <atomic>
#include <cmath>
#include <vector>

#include "effect.h"
#include "registry.h"

// short lfo-swept delay with feedback. the moving delay line, mixed back with the
// dry signal, sweeps a comb filter across the spectrum for the whoosh.
class Flanger : public Effect {
public:
  Flanger() {
    param("rate", &rate, 0.05f, 5.0f);
    param("depth", &depth, 0.0f, 1.0f);
    param("feedback", &feedback, 0.0f, 0.9f);
    param("mix", &mix, 0.0f, 1.0f);
  }

  void prepare(double sampleRate, int) override {
    sr_ = (float)sampleRate;
    buf_.assign((size_t)(0.03f * sr_) + 2, 0.0f);  // 30 ms line
    idx_ = 0;
    phase_ = 0.0f;
  }

  void process(float* buffer, int numFrames) override {
    float rt = rate.load(std::memory_order_relaxed);
    float dp = depth.load(std::memory_order_relaxed);
    float fb = feedback.load(std::memory_order_relaxed);
    float mx = mix.load(std::memory_order_relaxed);
    int n = (int)buf_.size();
    float base = 0.001f * sr_;       // 1 ms base delay
    float sweep = 0.005f * sr_ * dp;  // up to 5 ms of sweep
    float inc = 2.0f * 3.14159265f * rt / sr_;
    for (int i = 0; i < numFrames; i++) {
      float lfo = 0.5f * (1.0f + std::sin(phase_));
      phase_ += inc;
      if (phase_ > 6.2831853f) phase_ -= 6.2831853f;

      float ds = base + sweep * lfo;
      float rp = (float)idx_ - ds;
      while (rp < 0.0f) rp += n;
      int r0 = (int)rp;
      int r1 = r0 + 1 >= n ? 0 : r0 + 1;
      float frac = rp - (float)r0;
      float delayed = buf_[r0] * (1.0f - frac) + buf_[r1] * frac;

      buf_[idx_] = buffer[i] + delayed * fb;
      buffer[i] = buffer[i] * (1.0f - mx) + delayed * mx;
      if (++idx_ >= n) idx_ = 0;
    }
  }
  const char* name() const override { return "flanger"; }

private:
  std::atomic<float> rate{0.3f};
  std::atomic<float> depth{0.7f};
  std::atomic<float> feedback{0.3f};
  std::atomic<float> mix{0.5f};
  std::vector<float> buf_;
  int idx_ = 0;
  float phase_ = 0.0f;
  float sr_ = 48000.0f;
};

REGISTER_EFFECT(Flanger, "flanger")

File: effects/flanger.lua
-- short lfo-swept delay with feedback. the moving delay line, mixed back with
-- the dry signal, sweeps a comb filter across the spectrum for the whoosh.
name = "flanger"
params = {
  { name = "rate",     min = 0.05, max = 5,   default = 0.3 },
  { name = "depth",    min = 0,    max = 1,   default = 0.7 },
  { name = "feedback", min = 0,    max = 0.9, default = 0.3 },
  { name = "mix",      min = 0,    max = 1,   default = 0.5 },
}

local ffi = require("ffi")
local floor, sin = math.floor, math.sin
local tau = 2 * math.pi

local sr, size, buf, idx, phase

function prepare(sampleRate, maxBlock)
  sr = sampleRate
  size = floor(0.03 * sr) + 2
  buf = ffi.new("float[?]", size)
  idx = 0
  phase = 0
end

prepare(48000, 0)

function process(b, n, p)
  local base = 0.001 * sr
  local sweep = 0.005 * sr * p.depth
  local inc = tau * p.rate / sr
  local fb, mx = p.feedback, p.mix
  for i = 0, n - 1 do
    local lfo = 0.5 * (1 + sin(phase))
    phase = phase + inc
    if phase > tau then phase = phase - tau end

    local ds = base + sweep * lfo
    local rp = idx - ds
    while rp < 0 do rp = rp + size end
    local r0 = floor(rp)
    local r1 = r0 + 1
    if r1 >= size then r1 = 0 end
    local frac = rp - r0
    local delayed = buf[r0] * (1 - frac) + buf[r1] * frac

    buf[idx] = b[i] + delayed * fb
    b[i] = b[i] * (1 - mx) + delayed * mx
    idx = idx + 1
    if idx >= size then idx = 0 end
  end
end
